service/src/main/java/bio/terra/workspace/common/utils/GcpUtils.java

```diff
@@ -82,6 +83,9 @@ public static void pollAndRetry(
         // do not waste time retrying on client error.
         throw new BadRequestException(
             String.format("Gcp calls failed with client error code %s. Do not retry", code));
+      } else if (code == Code.PERMISSION_DENIED.value()) {
```

Comment:
Is permission_denied 403? how do we even get to here and skip the is4xxClientError check? 

also surprising that it throws permission_denied for quota limit. could you add a comment here explaining that this also handles quota limit execeed error?

Reply:
Yes, it's 403. So, the `is4xxClientError` check has the following [subset](https://github.com/DataBiosphere/terra-workspace-manager/blob/main/service/src/main/java/bio/terra/workspace/common/utils/GcpUtils.java#L103) of codes:

```java
private static boolean is4xxClientError(int code) {
    return Code.INVALID_ARGUMENT.value() == code // 400
        || Code.OUT_OF_RANGE.value() == code // 400
        || Code.FAILED_PRECONDITION.value() == code // 400
        || Code.ALREADY_EXISTS.value() == code // 409
        || Code.ABORTED.value() == code // 409
        || Code.UNAUTHENTICATED.value() == code // 401
        || Code.RESOURCE_EXHAUSTED.value() == code; // 429
  }
```

I think the `PERMISSION_DENIED` refers to *not being allowed* to create *because* of the quota limit.

I'll add a comment about the quota limit!

Reply:
i'm personally ok with putting it in the is4xxClientError and just call it a badRequest too... since we are not throwing conflict for 409 or forbidden for 401.